extension-manifest-v3/src/background/utils/setup.js

```diff
@@ -0,0 +1,32 @@
+/**
+ * Ghostery Browser Extension
+ * https://www.ghostery.com/
+ *
+ * Copyright 2017-present Ghostery GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0
+ */
+
+export default function asyncSetup(promises, threshold = 5000) {
+  let timeoutId;
+
+  const result = {
+    pending: Promise.race([
+      Promise.all(promises),
+      new Promise((_, reject) => {
+        timeoutId = setTimeout(() => {
+          reject(Error('Initial setup threshold exceeded'));
+        }, threshold);
+      }),
+    ]).then((...args) => {
+      result.pending = false;
+      clearTimeout(timeoutId);
+
+      return args;
+    }),
+  };
+
```

Comment:
I would suggest to have a catcher statement if the promise fails (timeout or error) and warns.

```
  result.pending.catch((e) => {
    console.warn('Error during setup', e);
  });
  ```
  
  Then you get a proper stacktrace that points to the original caller:
  
  
![better-error](https://github.com/ghostery/ghostery-extension/assets/1616509/24cd6479-45a6-4cd7-a822-b1005515d18e)

It also avoids a rejected promise warning, since for instance, in reporting only the handler would eventually handle the message. But if there are no event, it will not run quickly enough and Chrome thinks it is unrejected and gives you a potentially misleading error.